domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -191,6 +245,215 @@ class TopicController @Inject constructor(
+      AsyncResult.success(completedStoryListBuilder.build())
+    } else {
+      AsyncResult.failed(CompletedStoryListNotFoundException("ProfileId $profileId contain any completed story list."))
+    }
+  }
+
+  /** Combines the specified topic without progress and topic-progress into a topic. */
+  private fun combineTopicAndTopicProgress(topic: Topic, topicProgress: TopicProgress): Topic {
+    val topicBuilder = topic.toBuilder()
+    if (topicProgress.storyProgressMap.isNotEmpty()) {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
+        val storyBuilder = storySummary.toBuilder()
+        if (topicProgress.storyProgressMap.containsKey(storySummary.storyId)) {
+          storySummary.chapterList.forEachIndexed { chapterIndex, chapterSummary ->
+            if (topicProgress.storyProgressMap[storySummary.storyId]!!.chapterProgressMap.containsKey(chapterSummary.explorationId)) {
+              val chapterBuilder = chapterSummary.toBuilder()
+              chapterBuilder.chapterPlayState = ChapterPlayState.COMPLETED
+              storyBuilder.setChapter(chapterIndex, chapterBuilder)
+            } else {
+              if (storyBuilder.getChapter(chapterIndex - 1).chapterPlayState == ChapterPlayState.COMPLETED) {
+                val chapterBuilder = chapterSummary.toBuilder()
+                chapterBuilder.chapterPlayState = ChapterPlayState.NOT_STARTED
+                storyBuilder.setChapter(chapterIndex, chapterBuilder)
+              } else {
+                val chapterBuilder = chapterSummary.toBuilder()
+                chapterBuilder.chapterPlayState = ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES
+                storyBuilder.setChapter(chapterIndex, chapterBuilder)
+              }
+            }
+          }
+          topicBuilder.setStory(storyIndex, storyBuilder.build())
+        } else {
+          if (storySummary.chapterList.isNotEmpty()) {
+            val updatedStorySummary = setFirstChapterAsNotStarted(storySummary)
+            topicBuilder.setStory(storyIndex, updatedStorySummary)
+          }
+        }
+      }
+    } else {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
+        if (storySummary.chapterList.isNotEmpty()) {
+          val updatedStorySummary = setFirstChapterAsNotStarted(storySummary)
+          topicBuilder.setStory(storyIndex, updatedStorySummary)
+        }
+      }
+    }
+    return topicBuilder.build()
+  }
+
+  /** Combines the specified story-summary without progress and story-progress into a new topic. */
+  private fun combineStorySummaryAndStoryProgress(
+    storySummary: StorySummary,
+    storyProgress: StoryProgress
+  ): StorySummary {
+    if (storyProgress.chapterProgressMap.isNotEmpty()) {
+      val storyBuilder = storySummary.toBuilder()
+      storySummary.chapterList.forEachIndexed { chapterIndex, chapterSummary ->
+        if (storyProgress.chapterProgressMap.containsKey(chapterSummary.explorationId)) {
```

Comment:
Here & below: this code seems duplicative with the logic for computing chapter progress above. Can we consolidate it? Prefer to de-duplicate.

Reply:
Using `combineStorySummaryAndStoryProgress` in `combineTopicAndTopicProgress` to de-duplicate the logic of `combineStorySummaryAndStoryProgress`

Reply:
Much simpler—thanks!